File: openInterface/microbit/assets/lib/microbit/stepper.py
from utime import sleep_ms

MAX_STEPS = 2048  # Nombre de pas pour 1 tour (à ajuster selon le moteur/driver)

class StepperMotor:

    STEPS = 0
    ROTATIONS = 1
    
    def __init__(self, in1, in2, in3, in4):
        self.input1 = in1
        self.input2 = in2
        self.input3 = in3
        self.input4 = in4
        self.delay_ms = 3  # Délai entre chaque pas (en ms)
        self.state = 0
    
    def setState(self, state):
        self.state = state

    def setDelay(self, delay_ms):
        self.delay_ms = delay_ms

    def steps(self, direction):
        if (self.state == 0):
            self.input1.write_digital(0)
            self.input2.write_digital(0)
            self.input3.write_digital(0)
            self.input4.write_digital(0)
        elif (self.state == 1):
            self.input1.write_digital(1)
            self.input2.write_digital(0)
            self.input3.write_digital(0)
            self.input4.write_digital(1)
        elif (self.state == 2):
            self.input1.write_digital(0)
            self.input2.write_digital(0)
            self.input3.write_digital(1)
            self.input4.write_digital(1)
        elif (self.state == 3):
            self.input1.write_digital(0)
            self.input2.write_digital(1)
            self.input3.write_digital(1)
            self.input4.write_digital(0)
        elif (self.state == 4):
            self.input1.write_digital(1)
            self.input2.write_digital(1)
            self.input3.write_digital(0)
            self.input4.write_digital(0)
        self.state += direction
        if (self.state < 1):
            self.state = 4
        elif (self.state > 4):        
            self.state = 1
    
    def moveAntiClockwise(self, steps, unit):
        if (unit == self.ROTATIONS):
            steps = steps * MAX_STEPS
        elif (unit == self.STEPS):
            steps = steps
        for _ in range(steps):
            self.steps(1)
            sleep_ms(self.delay_ms)
        self.state = 0

    def moveClockwise(self, steps, unit):
        if (unit == self.ROTATIONS):
            steps = steps * MAX_STEPS
        elif (unit == self.STEPS):
            steps = steps
        for _ in range(steps):
            self.steps(-1)
            sleep_ms(self.delay_ms)
        self.state = 0
    
    def stop(self):
        self.state = 0

File: openInterface/microbit/assets/js/simulator/src/micropython/stepper.js
// micro:bit - stepper module

const $builtinmodule = function (name) {

    const stepper = {};
    stepper.__name__ = new Sk.builtin.str("stepper");

    // --- Contrôleur anti-chevauchement : par moteur ---
    const GLOBAL_KEY = "__stepperMotorCtrl__";
    const root = (typeof window !== "undefined") ? window : globalThis;
    if (!root[GLOBAL_KEY]) {
        root[GLOBAL_KEY] = { gens: Object.create(null) }; // { [motorId]: genNumber }
    }
    function nextGenFor(motorId) {
        const g = root[GLOBAL_KEY].gens;
        g[motorId] = (g[motorId] || 0) + 1;
        return g[motorId];
    }
    function stillActive(motorId, myGen) {
        return root[GLOBAL_KEY].gens[motorId] === myGen;
    }

    const StepperMotor = function ($gbl, $loc) {

        $loc.STEPS = new Sk.builtin.int_(0);
        $loc.ROTATIONS = new Sk.builtin.int_(1);
        const MAX_STEPS = 2048;

        StepperMotor__init__ = function (self, in1, in2, in3, in4) {
            Sk.builtin.pyCheckArgsLen("__init__", arguments.length, 5, 5);
            self.in1 = in1;
            self.in2 = in2;
            self.in3 = in3;
            self.in4 = in4;
            self.step = 0;
            self.delay_ms = 3;

            self.stepperMotorId = "#stepper-motor_" + self.in1.name;
            // (optionnel) stocker la dernière promesse si tu veux un join plus tard
            self._lastPromise = null;

            let subtitleElement = $(self.stepperMotorId).find(".subtitle-module");
            if (subtitleElement.length === 0) {
                $(self.stepperMotorId).find(".module-header").append("<br>");
                subtitleElement = $("<span>").addClass("subtitle-module");
                $(self.stepperMotorId).find(".module-header").append(subtitleElement);
            }
            subtitleElement.html("P" + self.in1.name + "/P" + self.in2.name + "/P" + self.in3.name + "/P" + self.in4.name);
        };
        StepperMotor__init__.co_varnames = ['self', 'in1', 'in2', 'in3', 'in4'];
        $loc.__init__ = new Sk.builtin.func(StepperMotor__init__);

        const moveClockwise = function (self, steps, unit) {
    let steps_ = Sk.ffi.remapToJs(steps);
    const unit_ = Sk.ffi.remapToJs(unit);
    const MAX_STEPS = 2048;
    if (unit_ === 1) steps_ = steps_ * MAX_STEPS;

    const myGen = nextGenFor(self.stepperMotorId);

    // fire-and-forget
    self._lastPromise = (async () => {
        for (let i = 0; i < steps_; i++) {
            if (!stillActive(self.stepperMotorId, myGen)) return; // annulé par un nouveau move sur ce moteur
            self.step++;
            $(self.stepperMotorId + "_anim").css({ WebkitTransform: 'rotate(' + self.step * (360 / MAX_STEPS) + 'deg)' });
            $(self.stepperMotorId + "_value").html(self.step);
            await sleep_ms(self.delay_ms);
        }
        if (stillActive(self.stepperMotorId, myGen)) {
            $(self.stepperMotorId + "_anim").css('animation', 'unset');
        }
    })();

    // non bloquant -> retourner None immédiatement
    return Sk.builtin.none.none$;
};
$loc.moveClockwise = new Sk.builtin.func(moveClockwise);

const moveAntiClockwise = function (self, steps, unit) {
    let steps_ = Sk.ffi.remapToJs(steps);
    const unit_ = Sk.ffi.remapToJs(unit);
    const MAX_STEPS = 2048;
    if (unit_ === 1) steps_ = steps_ * MAX_STEPS;

    const myGen = nextGenFor(self.stepperMotorId);

    self._lastPromise = (async () => {
        for (let i = 0; i < steps_; i++) {
            if (!stillActive(self.stepperMotorId, myGen)) return;
            self.step--;
            $(self.stepperMotorId + "_anim").css({ WebkitTransform: 'rotate(' + self.step * (360 / MAX_STEPS) + 'deg)' });
            $(self.stepperMotorId + "_value").html(self.step);
            await sleep_ms(self.delay_ms);
        }
        if (stillActive(self.stepperMotorId, myGen)) {
            $(self.stepperMotorId + "_anim").css('animation', 'unset');
        }
    })();

    return Sk.builtin.none.none$;
};
$loc.moveAntiClockwise = new Sk.builtin.func(moveAntiClockwise);

        const setDelay = function (self, delay_ms) {
            self.delay_ms = Sk.ffi.remapToJs(delay_ms);
        };
        $loc.setDelay = new Sk.builtin.func(setDelay);

        // (Optionnel) Méthode stop si tu veux arrêter immédiatement un moteur
        const stop = function (self) {
            // invalide la génération courante de CE moteur
            nextGenFor(self.stepperMotorId);
        };
        $loc.stop = new Sk.builtin.func(stop);

    };

    stepper.StepperMotor = new Sk.misceval.buildClass(stepper, StepperMotor, "StepperMotor", []);
    return stepper;
};